### DISCUSSION TITLE
Remove (3D)

### DISCUSSION THREAD

**Discussion by jgomez720** [KittyCAD/modeling-app#970] [New Feature Requests]
### Description
All extrudes should be able to remove material as well.

### Functionality
* Have an option in Extrude to remove material. The default behaviour of extrude should be to add material, but the user should have an option to remove it.
* Shouldn't work if the user is trying to extrude in the opposite direction of correct model body.

### Supporting Docs/Images

**Extrusion Dialogue**
<img width="1582" alt="Screenshot 2023-11-01 at 10 17 45 AM" src="https://github.com/KittyCAD/modeling-app/assets/114548659/ad2024e4-f21f-4b2b-865a-ff42cd252c96">

**Error message if user tries to remove in opposite direction of model body**
<img width="1582" alt="Screenshot 2023-11-01 at 10 26 47 AM" src="https://github.com/KittyCAD/modeling-app/assets/114548659/a4e95ceb-999b-4ddb-82ee-bd0906319680">

**Extrusion with (remove) selected**
<img width="1582" alt="Screenshot 2023-11-01 at 10 18 15 AM" src="https://github.com/KittyCAD/modeling-app/assets/114548659/4e49f3b7-788c-481a-95f2-756a319d897d">

**End Result**
<img width="1582" alt="Screenshot 2023-11-01 at 10 18 22 AM" src="https://github.com/KittyCAD/modeling-app/assets/114548659/a6302469-bead-417c-9dd9-11db0865aad2">

## PseudoKCL

the `|> extrude()` function adds material by default, but will expect the parameter `remove: true`.
```
const box = startSketchOn('XY')
  |> startProfileAt([-9, 6], %)
  |> line([18, 0], %)
  |> line([0, -12], %)
  |> line([-18, 0], %)
  |> close(%)
  |> extrude(4, %)

const extrudeRemove = startSketchOn(ExtrusionStart(box))
  |> startProfileAt([-6, 3])
  |> line([12, 0], %)
  |> line([0, -6], %)
  |> line([-12, 0], %)
  |> extrude(1, remove: true, %)
```

**Expected Result**
![Screenshot 2023-10-18 at 2 30 54 PM](https://github.com/KittyCAD/modeling-app/assets/114548659/ef08c167-d91c-4b57-80e5-ee24700e3d58)

**Comment by adamchalmers** [comment]
Why not just use a normal boolean operation?

```
const outer = startSketchOn('XY')
  |> startProfileAt([-9, 6], %)
  |> line([18, 0], %)
  |> line([0, -12], %)
  |> line([-18, 0], %)
  |> close(%)
  |> extrude(4, %)

const inner = startSketchOn('XY')
  |> startProfileAt([-6, 3])
  |> line([12, 0], %)
  |> line([0, -6], %)
  |> line([-12, 0], %)
  |> extrude(1, %)

const finalShape = outer.remove(inner)
```

**Reply by alteous** [reply]
Has this been implemented in the engine?

**Reply by adamchalmers** [reply]
No, not yet.

To answer my own question above: we shouldn't use general-purpose subtraction. Here's why.

Kurt and Josh pointed out that remove-extrude is a special case of more general 3D constructive geometry subtraction. The former is much easier than the latter (as we discussed in-person with Alyn, fully general 3D subtraction is, mathematically speaking, Fucking Hard). So we should probably prefer a dedicated remove-extrude operation because fully general 3D subtract will be a while away.

**Comment by jgomez720** [comment]
oh I think David was commenting because this discussion got closed by mistake. I reopened it
